### ISSUE TITLE
TwitterRequestHandler QueryCompleted method and RateLimitUpdater issue

### ISSUE THREAD

**Issue by iandillon** [linvi/tweetinvi#1077]
Hi I am hacking around providing a solution to handling rate limits for v2 search so please forgive me if it seems to you to be a false report but I thought this worth noting as something you would be able to understand much better than I if there is a  problem in TweetInvi which needs resolving.

Within TwitterRequestHandler the QueryCompleted method is invoked and the Tweetinvi logic intends the QueryExecutedAsync method of RateLimitUpdater to be invoked within that method. I found that this was not the case and therefore information from rate limit headers was not updating an endpoint rate limit.

This may not be surfaced as other code refreshing credential rate limits queries the rate limit status endpoint and seems to get the updated values there.

Whilst debugging this seemed to arise as a result of the QueryCompleted method not being marked as async and line 157 rateLimitUpdater.QueryExecutedAsync(request.Query.Url, request.Query.TwitterCredentials, rateLimitHeaders); not being awaited.

Changing void QueryCompleted to 'async Task QueryCompleted' and rateLimitUpdater.QueryExecutedAsync to 'await rateLimitUpdater.QueryExecutedAsync' seemed to resolve the problem

If correct this problem also seems valid for void HandleException which maybe should be 'async Task HandleException' with the call to rateLimitUpdater.ClearRateLimitsForQueryAsync awaited.

Thanks

**Comment by linvi** [comment]
Thank you for your time and efforts looking at this issue. I have found the issue you are referring to and I am releasing both a fix and 2 improvements for RateLimits.

1. Unknown RateLimits will be automatically added when detected in the headers.
2. Unknown RateLimits will be added to the event : `TweetinviEvents.AfterExecutingRequest` right after the request has been executed.
3. You can request the rateLimits manually for unknown rate limits as followed:

``` c#
var rateLimits = await userClient.RateLimits.GetRateLimitsAsync();
var searchV2RateLimit = rateLimits.OtherEndpointRateLimits.FirstOrDefault(x => x.Key.EndpointURL == "https://api.twitter.com/2/tweets/search/recent").Value
```

I understand this can be improved more but I hope this will help you resolving your problem.

**Comment by linvi** [comment]
I have released a new version Tweetinvi 5.0.2 which fixes this issue and provide the improvements mentioned above.

Thank you for taking the time to report it.
Cheers,
Linvi
